backend/discord.py

```diff
@@ -32,10 +37,159 @@ async def fetch_bearer_token(code: str, redirect: str, *, refresh: bool) -> dict
 
 async def fetch_user_details(bearer_token: str) -> dict:
     async with httpx.AsyncClient() as client:
-        r = await client.get(f"{DISCORD_API_BASE_URL}/users/@me", headers={
+        r = await client.get(f"{constants.DISCORD_API_BASE_URL}/users/@me", headers={
             "Authorization": f"Bearer {bearer_token}"
         })
 
         r.raise_for_status()
 
         return r.json()
+
+
+async def _get_role_info() -> list[models.DiscordRole]:
+    """Get information about the roles in the configured guild."""
+    async with httpx.AsyncClient() as client:
+        r = await client.get(
+            f"{constants.DISCORD_API_BASE_URL}/guilds/{constants.DISCORD_GUILD}/roles",
+            headers={"Authorization": f"Bot {constants.DISCORD_BOT_TOKEN}"}
+        )
+
+        r.raise_for_status()
+        return [models.DiscordRole(**role) for role in r.json()]
+
+
+async def get_roles(
+    database: Database, *, force_refresh: bool = False
+) -> list[models.DiscordRole]:
+    """
+    Get a list of all roles from the cache, or discord API if not available.
+
+    If `force_refresh` is True, the cache is skipped and the roles are updated.
```

Comment:
I feel like this doesn't reflect the fact that the entire cache appears to be cleared?

Reply:
Does it... matter? We clear the entire cache because discord returns all the roles in one request, so clearing and adding everything is the most efficient course of action.

Reply:
I see, right. I think I read that sentence differently and focused on the cache being skipped.

This makes sense though, I'll mark this as resolved.